snafu-derive/src/lib.rs

```diff
@@ -985,18 +1020,35 @@ impl syn::parse::Parse for DocComment {
     }
 }
 
-fn attributes_from_syn(attrs: Vec<syn::Attribute>) -> MultiSynResult<Vec<SnafuAttribute>> {
+/// Given a list of syn `Attribute`, parses each attribute into a type known to SNAFU, namely a
+/// variant of SnafuAttribute.  Returns a list of pairs, where each pair is a SnafuAttribute and
+/// the TokenStream where it was found; the TokenStream can be used to give an accurate error
+/// message if there's a problem with the attribute.
+fn attributes_from_syn(
+    attrs: &[syn::Attribute],
+) -> MultiSynResult<Vec<(SnafuAttribute, proc_macro2::TokenStream)>> {
```

Comment:
What do you think about putting the `TokenStream` inside the `SnafuAttribute` itself, instead of returning the tuple?

Reply:
I had considered that, and it seemed like the best approach would be to replace `SnafuAttribute` with a struct that carries the location and the enum currently called `SnafuAttribute`, rather than putting the location in each variant of the current enum.  In the end I went with the modification to `attributes_from_syn` because the returned tuples felt natural to me, and it didn't require modifying every use of `SnafuAttribute` to add a field indirection.  I don't feel strongly, though.  What do you think is best of those?  Or do you have other implementation ideas that would simplify storing location with the `SnafuAttribute`?

Reply:
> replace `SnafuAttribute` with a struct that carries the location and the enum currently called `SnafuAttribute`

Yeah, the pattern of "an enum where every variant has the same piece of data" is a rather annoying case. I flip between the solution you mentioned and

```rust
enum E {
    A(Same, Diff1),
    B(Same, Diff2,
    C(Same, Diff3),
}

impl E {
    fn same(&self) -> &Same { /* ... */ }
}
```

I'm OK with the tuples for now, but I have the feeling that if it grows that the pieces will want names, growing into the enum-wrapped-struct you mentioned.